c/src/nnstreamer-capi-pipeline.c

```diff
@@ -1566,14 +1567,126 @@ ml_pipeline_src_input_data (ml_pipeline_src_h h, ml_tensors_data_h data,
   handle_exit (h);
 }
 
+/**
+ * @brief Internal function for appsrc callback - need_data.
+ */
+static void
+_pipe_src_cb_need_data (GstAppSrc * src, guint length, gpointer user_data)
+{
+  ml_pipeline_element *elem;
+  ml_pipeline_common_elem *src_h;
+  ml_pipeline_src_callbacks *src_cb;
+  GList *l;
+
+  elem = (ml_pipeline_element *) user_data;
+  g_mutex_lock (&elem->lock);
+
+  /* Iterate e->handles */
+  for (l = elem->handles; l != NULL; l = l->next) {
+    src_h = l->data;
+    if (src_h->callback_info == NULL)
+      continue;
+
+    src_cb = &src_h->callback_info->src_cb;
+    if (src_cb->need_data)
+      src_cb->need_data (src_h, length, src_h->callback_info->pdata);
+  }
+
+  g_mutex_unlock (&elem->lock);
+}
+
+/**
+ * @brief Internal function for appsrc callback - enough_data.
+ */
+static void
+_pipe_src_cb_enough_data (GstAppSrc * src, gpointer user_data)
+{
+  ml_pipeline_element *elem;
+  ml_pipeline_common_elem *src_h;
+  ml_pipeline_src_callbacks *src_cb;
+  GList *l;
+
+  elem = (ml_pipeline_element *) user_data;
+  g_mutex_lock (&elem->lock);
+
+  /* Iterate e->handles */
+  for (l = elem->handles; l != NULL; l = l->next) {
+    src_h = l->data;
+    if (src_h->callback_info == NULL)
+      continue;
+
+    src_cb = &src_h->callback_info->src_cb;
+    if (src_cb->enough_data)
+      src_cb->enough_data (src_h, src_h->callback_info->pdata);
+  }
+
+  g_mutex_unlock (&elem->lock);
+}
+
+/**
+ * @brief Internal function for appsrc callback - seek_data.
+ */
+static gboolean
+_pipe_src_cb_seek_data (GstAppSrc * src, guint64 offset, gpointer user_data)
+{
+  ml_pipeline_element *elem;
+  ml_pipeline_common_elem *src_h;
+  ml_pipeline_src_callbacks *src_cb;
+  GList *l;
+
+  elem = (ml_pipeline_element *) user_data;
+  g_mutex_lock (&elem->lock);
+
+  /* Iterate e->handles */
+  for (l = elem->handles; l != NULL; l = l->next) {
+    src_h = l->data;
+    if (src_h->callback_info == NULL)
+      continue;
+
+    src_cb = &src_h->callback_info->src_cb;
+    if (src_cb->seek_data)
+      src_cb->seek_data (src_h, offset, src_h->callback_info->pdata);
+  }
+
+  g_mutex_unlock (&elem->lock);
+  return TRUE;
+}
+
 /**
  * @brief Register callbacks for src events (more info in nnstreamer.h)
  */
 int
 ml_pipeline_src_input_callback (ml_pipeline_src_h src_handle,
-    ml_pipeline_src_callbacks cb, void *user_data)
+    ml_pipeline_src_callbacks * cb, void *user_data)
 {
-  return ML_ERROR_NOT_SUPPORTED; /** @todo NYI */
+  GstAppSrcCallbacks appsrc_cb = { 0, };
```

Comment:
Is it supposed to have a single handle for sink element? or are you going to support MULTIPLE handles for a sink element?

Reply:
Having multiple handlers for a sink event appears very problematic.

Reply:
I removed iteration. Thanks.